### PR TITLE
feat: attributes

### LINKED ISSUE #12738: [CAL-2793] Attributes in teams
Prepare the Attributes project

for "marketplace embed" and for routing forms

[**Designs**](https://www.figma.com/file/UajFu4M1APhn2AywAEJkJr/New-Features?type=design&node-id=15496%3A500553&mode=dev)

<sub>From [SyncLinear.com](https://synclinear.com) | [CAL-2793](https://linear.app/calcom/issue/CAL-2793/attributes-in-teams)</sub>

### PR DESCRIPTION
**Pull request by sean-brydon**
## What does this PR do?

Fixes: #12738 

Loom: https://www.loom.com/share/f2b28f154bb040c7805844fad019f790

**How to test:**
1) Create an organization
2) Add a member to the organization
3) go to attributes in the organization nav bar
4) create a few attributes of different types
5) goto users page 
6) Edit>assign attributes - save.
7) open the slide over in viewing mode - see attributes saved

### CODE REVIEW

File: apps/api/v2/src/ee/event-types/event-types_2024_04_15/controllers/event-types.controller.ts
Comment (sean-brydon) [inline apps/api/v2/src/ee/event-types/event-types_2024_04_15/controllers/event-types.controller.ts:43]: Prettier fixing this auto for me

File: packages/app-store/skype/config.json
Comment (sean-brydon) [inline packages/app-store/skype/config.json:1]: Prettier/eeslint fixing this on commit - not 100% sure

File: packages/features/ee/organizations/pages/settings/attributes/AttributesForm.tsx
Comment (sean-brydon) [inline packages/features/ee/organizations/pages/settings/attributes/AttributesForm.tsx:106]: Show "more" options if we're on select or multiselect

File: packages/features/users/components/UserTable/EditSheet/EditUserForm.tsx
Comment (sean-brydon) [inline packages/features/users/components/UserTable/EditSheet/EditUserForm.tsx:298]: Merges users attributes from seperate endpoint with all org attributes to create default values

File: packages/features/ee/organizations/pages/settings/attributes/attributes-list-view.tsx
Comment (CarinaWolli) [inline packages/features/ee/organizations/pages/settings/attributes/attributes-list-view.tsx:116]: should we have a modal to confirm deletion?

**Review by CarinaWolli — CHANGES_REQUESTED** [review]
Nice work! I left you some comments, some can probably handled in a follow-up. Requesting changes until my comments are addressed 🙏

File: packages/features/ee/organizations/pages/settings/attributes/AttributesForm.tsx
Comment (joeauyeung) [inline packages/features/ee/organizations/pages/settings/attributes/AttributesForm.tsx:96]: Why do we want to disable deleting the first option? If I have multiple options, what if I wanted to delete the first option?

Should we instead disable it if it's the only option?

File: packages/features/users/components/UserTable/EditSheet/EditUserSheet.tsx
Comment (joeauyeung) [inline packages/features/users/components/UserTable/EditSheet/EditUserSheet.tsx:1]: I got an error saying the org is not reviewed but I was still able to save the attributes against the user.

File: packages/trpc/server/routers/viewer/attributes/assignUserToAttribute.handler.ts
Comment (joeauyeung) [inline packages/trpc/server/routers/viewer/attributes/assignUserToAttribute.handler.ts:109]: I noticed in this map we're making a lot of DB queries and writes. Instead could we make a singular DB call to get the attributes associated with the user and then a single `updateMany` call after iterating over `filteredAttributes`?

File: packages/trpc/server/routers/viewer/attributes/assignUserToAttribute.handler.ts
Comment (joeauyeung) [inline packages/trpc/server/routers/viewer/attributes/assignUserToAttribute.handler.ts:16]: Could we use the `authedOrgAdminProcedure` in the `_router`?

File: packages/trpc/server/routers/viewer/attributes/getByUserId.handler.ts
Comment (joeauyeung) [inline packages/trpc/server/routers/viewer/attributes/getByUserId.handler.ts:27]:
```suggestion
const getByUserIdHandler = async ({ input, ctx }: GetOptions) => {
```

File: packages/trpc/server/routers/viewer/attributes/getByUserId.handler.ts
Comment (joeauyeung) [inline packages/trpc/server/routers/viewer/attributes/getByUserId.handler.ts:37]:
```suggestion
  // Ensure user is apart of the organization
```

**Review by joeauyeung — CHANGES_REQUESTED** [review]
This is a huge undertaking @sean-brydon! I have some comments and thoughts 

## Blocking
- I noticed we have separate tRPC endpoints for attributes. Instead could we create an `AttributeRepository`? That way if a customer requests attribute API endpoints it would be easy to implement and we could write some tests against this new feature.

File: packages/lib/constants.ts
Comment (sean-brydon) [reply packages/lib/constants.ts:91]: Oh this was meant to be not commmited nice spot!

File: packages/features/settings/layouts/SettingsLayout.tsx
Comment (sean-brydon) [reply packages/features/settings/layouts/SettingsLayout.tsx:182]: It can be - but its a thing for all feature flags that are used in settings sidebar :/

File: packages/features/users/components/UserTable/EditSheet/EditUserForm.tsx
Comment (sean-brydon) [reply packages/features/users/components/UserTable/EditSheet/EditUserForm.tsx:306]: @CarinaWolli

File: packages/features/ee/organizations/pages/settings/attributes/attributes-create-view.tsx
Comment (sean-brydon) [reply packages/features/ee/organizations/pages/settings/attributes/attributes-create-view.tsx:1]: I was just matching designs - cc @ciaranha

File: packages/features/ee/organizations/pages/settings/attributes/attributes-list-view.tsx
Comment (sean-brydon) [reply packages/features/ee/organizations/pages/settings/attributes/attributes-list-view.tsx:154]: Ermm i don't see why we would to be honest - its not going to be used anywhere else

File: packages/features/users/components/UserTable/EditSheet/EditUserForm.tsx
Comment (sean-brydon) [reply packages/features/users/components/UserTable/EditSheet/EditUserForm.tsx:133]: I was thinking that a lot of this logic isnt really "user profile" related. Felt very out of place. 

I did this initaly but wasnt a fan. Let me know what you think - happy to move it back but it just adds more bloat to the update profile endpoint

File: packages/prisma/schema.prisma
Comment (sean-brydon) [reply packages/prisma/schema.prisma:1436]: No - i can change it

File: packages/trpc/server/routers/viewer/attributes/assignUserToAttribute.handler.ts
Comment (sean-brydon) [reply packages/trpc/server/routers/viewer/attributes/assignUserToAttribute.handler.ts:109]: I do agree with you here - but i battled and battled transactions - trying to do this in a cleaner way and everything i was trying was actually slower than this approach. 

If you have any ideas would love to try them

File: packages/features/users/components/UserTable/EditSheet/EditUserSheet.tsx
Comment (sean-brydon) [reply packages/features/users/components/UserTable/EditSheet/EditUserSheet.tsx:1]: This isnt even an error thrown here 🤔

File: packages/features/ee/organizations/pages/settings/attributes/attributes-list-view.tsx
Comment (sean-brydon) [reply packages/features/ee/organizations/pages/settings/attributes/attributes-list-view.tsx:116]: @ciaranha also here

File: packages/trpc/server/routers/viewer/attributes/assignUserToAttribute.handler.ts
Comment (sean-brydon) [reply packages/trpc/server/routers/viewer/attributes/assignUserToAttribute.handler.ts:97]: I just took a stab at this again - we can't really update any as all options for example have different values and not the same.

**Comment by sean-brydon** [comment]
> This is a huge undertaking @sean-brydon! I have some comments and thoughts
> 
> ## Blocking
> * I noticed we have separate tRPC endpoints for attributes. Instead could we create an `AttributeRepository`? That way if a customer requests attribute API endpoints it would be easy to implement and we could write some tests against this new feature.

I agree with you here - would you mind if we make that abstraction when i handle the api/v2 side of this. This PR is pretty big as it is and i'd much rather get this merged and do it in the next batch of this

File: packages/features/ee/organizations/pages/settings/attributes/attributes-create-view.tsx
Comment (ciaranha) [reply packages/features/ee/organizations/pages/settings/attributes/attributes-create-view.tsx:1]: We went with a page here because you could have quite a lot of options here:
CleanShot 2024-08-09 at 09 10 18@2x

I think one of the main things that feels off right now is that the header on the internal page doesn't line up with the header on the previous page and we switch between having a border and not having one so it really doesn't help with making the transition to a page feeling smooth.

File: packages/features/users/components/UserTable/EditSheet/EditUserForm.tsx
Comment (ciaranha) [reply packages/features/users/components/UserTable/EditSheet/EditUserForm.tsx]: Designs are updated.

File: packages/features/ee/organizations/pages/settings/attributes/attributes-list-view.tsx
Comment (ciaranha) [reply packages/features/ee/organizations/pages/settings/attributes/attributes-list-view.tsx:118]: Does that relate to this?

CleanShot 2024-08-09 at 09 15 05@2x

File: packages/features/ee/organizations/pages/settings/attributes/attributes-create-view.tsx
Comment (sean-brydon) [reply packages/features/ee/organizations/pages/settings/attributes/attributes-create-view.tsx:1]: For sure - lets make that transition between list view and this create/edit page seemless.

File: packages/features/ee/organizations/pages/settings/attributes/attributes-list-view.tsx
Comment (sean-brydon) [reply packages/features/ee/organizations/pages/settings/attributes/attributes-list-view.tsx:119]: Oh i missed this in designs lol - i can do this

File: packages/features/ee/organizations/pages/settings/attributes/attributes-create-view.tsx
Comment (sean-brydon) [reply packages/features/ee/organizations/pages/settings/attributes/attributes-create-view.tsx:1]: CleanShot 2024-08-09 at 10 02 37
Let me know what you think @ciaranha  @joeauyeung

File: packages/features/ee/organizations/pages/settings/attributes/attributes-list-view.tsx
Comment (sean-brydon) [reply packages/features/ee/organizations/pages/settings/attributes/attributes-list-view.tsx:119]: Pushed: 
CleanShot 2024-08-09 at 10 20 46

File: packages/features/ee/organizations/pages/settings/attributes/attributes-create-view.tsx
Comment (sean-brydon) [reply packages/features/ee/organizations/pages/settings/attributes/attributes-create-view.tsx:33]: Should be fixed now :)

File: packages/features/ee/organizations/pages/settings/attributes/attributes-create-view.tsx
Comment (joeauyeung) [reply packages/features/ee/organizations/pages/settings/attributes/attributes-create-view.tsx:41]: Can we add translations in the <Meta /> similar to here?
image

File: packages/features/users/components/UserTable/EditSheet/EditUserForm.tsx
Comment (joeauyeung) [reply packages/features/users/components/UserTable/EditSheet/EditUserForm.tsx:133]: I was thinking that the attributes logic would only run if they were changed. Also I was visioning the logic for editing attributes for a user would live under an `AttributesRepository` so it wouldn't add too much to the endpoint.

This still works so non-blocking.

File: packages/features/users/components/UserTable/EditSheet/EditUserSheet.tsx
Comment (joeauyeung) [reply packages/features/users/components/UserTable/EditSheet/EditUserSheet.tsx:1]: I wonder if this is because we have updating the user and updating the attributes in two separate endpoints. One passes but another fails.

https://www.loom.com/share/b7ba6d510f16422b94a1ffbf7cc651ae

**Review by joeauyeung — CHANGES_REQUESTED** [review]
Everything looks good. Just small nits but not blocking. The only thing that's stopping me from approving is deciding if the update user and update user attributes should be separate endpoints. I just found this that makes me lean towards having it all under one endpoint. 

https://github.com/calcom/cal.com/pull/15964#discussion_r1713951362

**Review by joeauyeung — DISMISSED** [review]
Saw all my comments have been addressed. I'll approve but let's wait until @CarinaWolli's comments have been addressed before merging.

**Comment by sean-brydon** [comment]
> Saw all my comments have been addressed. I'll approve but let's wait until @CarinaWolli's comments have been addressed before merging.

Should be all of them! Will fix merge conflicts right now

File: packages/features/ee/organizations/pages/settings/attributes/attributes-create-view.tsx
Comment (emrysal) [inline packages/features/ee/organizations/pages/settings/attributes/attributes-create-view.tsx:87]: Should be a translation I think

File: packages/features/ee/organizations/pages/settings/attributes/attributes-edit-view.tsx
Comment (CarinaWolli) [inline packages/features/ee/organizations/pages/settings/attributes/attributes-edit-view.tsx:35]: translation missing

File: packages/features/ee/organizations/pages/settings/attributes/attributes-edit-view.tsx
Comment (CarinaWolli) [inline packages/features/ee/organizations/pages/settings/attributes/attributes-edit-view.tsx:50]: Can we also add translations here?

**Review by CarinaWolli — DISMISSED** [review]
some translations are still missing, everything else looks good 🙏

**Review by zomars — APPROVED** [review]

**Pull request merged**
